Fix bug in Avro schema cleanup during cluster termination

vcrfxia opened the pull request:
### Description 

The `/ksql/terminate` REST endpoint offers the ability to cleanup sink topics and any associated Avro schemas specified by the request property `deleteTopicList`. However, we currently don't clean up sink schemas correctly if the ksqlTopicName (i.e., `KsqlTopic#topicName`) and kafkaTopicName (i.e., `KsqlTopic#kafkaTopicName`) for the sink topic don't match. The reason for this bug is because the subject name of the Avro schema is determined by the kafkaTopicName, but the cleanup code currently looks for schemas according to the ksqlTopicName. This PR fixes the bug.

### Testing done 

Updated unit tests. Added integration test.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

rodesai commented on ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:
```diff
@@ -84,7 +84,17 @@ public StructuredDataSource getSource(final String sourceName) {
+  public Optional<StructuredDataSource> getSourceForKafkaTopic(final String kafkaTopicName) {
+    return dataSources.values()
+        .stream()
+        .filter(p -> p.source.getKsqlTopic().getName() != null
```
what's this null check for? that's going to check the underlying ksql topic name (not the kafka topic name).

vcrfxia replied:
Fixed, thanks for catching that!

agavra commented on ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:
```diff
@@ -83,8 +83,18 @@ public StructuredDataSource getSource(final String sourceName) {
-        .filter(p -> p.source.getKsqlTopic().getName() != null
-            && p.source.getKsqlTopic().getName().equals(ksqlTopicName))
+        .filter(p -> p.source.getTopicName() != null
+            && p.source.getTopicName().equals(ksqlTopicName))
```
From the code it's pretty unclear what the difference between `getTopicName` and `getKafkaTopicName` are. I know it's not your code and it's probably out of scope for this PR but I think it'd be good to get a better naming scheme down so that future bugs around this are avoided. Perhaps something like renaming `KsqlTopic` to `KsqlTopicAlias`, `KsqlTopicReference` or something like that to avoid overloading "name".

hjafarpour replied:
I agree @agavra , if you remember our past discussion, we have KSQL topic which is defined for a Kafka topic. I think renaming `topicName` to `ksqlTopicName` and keeping `kafkaTopicName` would help reducing ambiguity here.

vcrfxia replied:
Completely agree, but want to keep this diff small since it's going into a branch past code freeze. Will do in a follow-up.

agavra commented on ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:
```diff
@@ -83,8 +83,18 @@ public StructuredDataSource getSource(final String sourceName) {
+  public Optional<StructuredDataSource> getSourceForKafkaTopic(final String kafkaTopicName) {
+    return dataSources.values()
+        .stream()
+        .filter(p -> p.source.getKafkaTopicName() != null
```
how is it possible for a `KsqlTopic` to have a `null` kafka topic name? Again out of scope for this PR, but unless I'm misunderstanding something we should probably push this null check into the constructor of `KsqlTopic`

vcrfxia replied:
It shouldn't be possible, but this isn't the right PR for it. Will look into it more in a follow-up.

agavra commented on ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:
```diff
@@ -83,8 +83,18 @@ public StructuredDataSource getSource(final String sourceName) {
+  @Override
+  public Optional<StructuredDataSource> getSourceForKafkaTopic(final String kafkaTopicName) {
+    return dataSources.values()
+        .stream()
```
nit: [not your original code] we can start by `.map(sourceInfo -> sourceInfo.source)` since we only use that anyway in the filter

vcrfxia replied:
Nice eye. Will do in a follow-up.

hjafarpour commented on ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStore.java:
```diff
@@ -28,6 +28,8 @@
 
   Optional<StructuredDataSource> getSourceForTopic(String ksqlTopicName);
 
+  Optional<StructuredDataSource> getSourceForKafkaTopic(String kafkaTopicName);
```
We have a one-to- many relation between kafka topic and StructuredDataSource. For instance, you can declare multiple KSQL streams on the same kafka topic. You need to return a collection(maybe set or list) here.

vcrfxia replied:
Good point. Switched to return a list instead.

hjafarpour commented on ksql-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java:
```diff
@@ -83,8 +83,18 @@ public StructuredDataSource getSource(final String sourceName) {
+  }
+
+  @Override
+  public Optional<StructuredDataSource> getSourceForKafkaTopic(final String kafkaTopicName) {
```
Same as the comment above, this should return a collection.

vcrfxia replied:
Done.

vcrfxia commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/ClusterTerminator.java:
```diff
@@ -116,17 +114,17 @@ private void deleteTopics(final List<String> topicsToBeDeleted) {
 
+  private boolean hasAvroSource(final String topicName) {
+    return ksqlEngine.getMetaStore().getSourcesForKafkaTopic(topicName).stream()
+        .anyMatch(dataSource -> dataSource.isSerdeFormat(DataSourceSerDe.AVRO));
```
This is a little overkill since if there are multiple streams defined on top of the same topic, they should all have the same format. Shouldn't hurt either though.

vcrfxia commented:
Thanks for the reviews @rodesai @spena @agavra and @hjafarpour . I've added an integration test around the behavior in this bug fix. If you could take another look, that'd be great. Thanks!

Note that I needed to introduce a new `buildApplication()` method to `KsqlRestApplication` so that the newly added integration test can inject a MockSchemaRegistryClient. In order to make this new `buildApplication()` method package-private, I moved TestKsqlRestApp from the `io.confluent.ksql.test.util` package into the `io.confluent.ksql.rest.server` package.

vcrfxia commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/ClusterTerminationTest.java:
```diff
@@ -0,0 +1,207 @@
+    TEST_HARNESS.verifySubjectAbsent(SINK_TOPIC + KsqlConstants.SCHEMA_REGISTRY_VALUE_SUFFIX);
+  }
+
+  private static List<KsqlEntity> awaitResults(final List<KsqlEntity> pending) {
```
This helper method and a bunch of the others below are duplicated here, and in RestApiTest and KsqlResourceFunctionalTest. I think they should be pulled out into a shared test util file, but would prefer making this change in a follow-up PR unless others feel strongly.

agavra commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/TestKsqlRestApp.java:
```diff
@@ -362,5 +388,9 @@ public Builder withProperties(final Map<String, ?> props) {
     }
+
+    public TestKsqlRestApp buildWithServiceContext(final Supplier<ServiceContext> serviceContext) {
+      return new TestKsqlRestApp(bootstrapServers, additionalProps, serviceContext);
```
nit: usually builder patterns would have `withServiceContext(Supplier<ServiceContext>)` - my preference would be to remove the old constructor and have just the new one, but (as with most style things) I'm happy to keep it as is if you feel strongly.

vcrfxia replied:
I considered that but thought it was weird to have a null-check in the `build()` method. Maybe you had something different in mind? I've implemented what I was thinking earlier, along the lines of your suggestion. Let me know what you think.

agavra replied:
My idea wasn't to have a null-check, but rather to have a non-final field in the builder with a default.
```java
private Supplier<ServiceContext> serviceContext = () -> defaultServiceContext(bootstrapServers, buildBaseConfig(additionalProps)));
```

vcrfxia replied:
But this default depends on `bootstrapServers` and `additionalProps`, which can change as a result of calls to `withProperty()` or `withProperties()`. In order to update the default correctly in these situations, we'd have to track whether `withServiceContext()` had been called or not, which is effectively the null-check currently implemented.

vcrfxia replied:
Resolved offline -- turns out setting the default once at the beginning works fine since `bootstrapServers` and `additionalProps` are final and their actual values won't be picked up until `.get()` is called on the supplier, which happens after `.build()`. Updated to use @agavra 's suggestion. Thanks!

vcrfxia commented on ksql-engine/src/test/java/io/confluent/ksql/integration/IntegrationTestHarness.java:
```diff
@@ -440,6 +445,82 @@ public void produceRecord(final String topicName, final String key, final String
+   *
+   * @param topicNames the names of the topics to check existence for.
+   */
+  public void verifyTopicsPresent(final String... topicNames) throws Exception {
```
Done.

vcrfxia commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/ClusterTerminationTest.java:
```diff
@@ -0,0 +1,207 @@
+  @Test
+  public void shouldCleanUpSinkTopicsAndSchemasDuringClusterTermination() throws Exception {
+    // Given:
+    makeKsqlRequest("CREATE STREAM " + SINK_STREAM
```
I'm not convinced of the value of having a test for the case where multiple sources are registered with a single topic, since people shouldn't be declaring multiple CSAS/CTAS statements with sink topics backed by the same topic. (If they wanted to achieve that, they should use CSAS/CTAS for the first statement and INSERT INTO for subsequent statements.) It seems plausible someone would want to declare multiple sources on top of the same topic via CREATE STREAM/TABLE statements, but you can't drop those from cluster termination, only sinks.

Also, there are a number of additional integration tests we should add around the cluster termination functionality, e.g., tests to verify that the command topic and processing log topics are cleaned up, and tests to verify that internal topics are cleaned up too. I'm holding off on adding those tests into this PR in order to keep the PR more targeted, but if people think there's value in getting those tests into 5.2, I can add them soon.

agavra replied:
> It seems plausible someone would want to declare multiple sources on top of the same topic via CREATE STREAM/TABLE statements, but you can't drop those from cluster termination, only sinks.

Ah yes, this makes a lot of sense (I was only referring to CT/CS)
